### ISSUE TITLE
Cannot use key and ciphers in different providers

### ISSUE THREAD

**Issue by selvanair** [openssl/openssl#16614]
For an external key (non-exportable) I use a custom provider with a minimal store, keymgmt and signature implemented. This works, the key can be loaded, and signing using EVP_PKEY_sign() calls the method in my provider. The latter requires the EVP_PKEY_CTX  created with the same propq as the key, though.

Now, to use the corresponding certificate in TLS, I create SSL_CTX in default context so that cipher and digest algorthms are available. And load the key as the private key. This errors during signature as it wants to export the key which is not possible. Is there a way out of this?

An edited snippet of the relevant code:
```
OSSL_PROVIDER_load(libctx, "ovpn.xkey");

EVP_PKEY *pkey = my_custom_loader(libctx, "provider=ovpn.xkey", uri); /* this loads an opaque pointer as keydata */ 

SSL_CTX *ctx = SSL_CTX_new_ex(NULL, NULL, method); /* using default context */
SSL_CTX_use_certificate(ctx, cert);
SSL_CTX_use_Privatekey(ctx, pkey);
..
```
In this case, all is good until  signature is needed which fails with calls to ovpn_keymgt_export() with PRIVATE_KEY in selection, not possible for the key. So, it looks like the sign op is tried in the context for SSL_CTX which needs the key exported. 

If I use `SSL_CTX_new_ex(libctx, "provider=ovpn.xkey", method)` it fails quickly with "library has no ciphers".  Is it possible to get the sign operation called in the context of the key's provider, without losing access to algorithms in the default provider?

**Comment by mattcaswell** [comment]
> If I use SSL_CTX_new_ex(libctx, "provider=ovpn.xkey", method) it fails quickly with "library has no ciphers"

Try:

````
SSL_CTX_new_ex(libctx, "?provider=ovpn.xkey", method)
````
This will prefer algorithms from your custom provider over algorithms from other providers. Of course you need to ensure that `libctx` also has the default provider loaded into it.

**Comment by selvanair** [comment]
Thanks a bunch. Much appreciated. Prefixing with "?" resolves cipher/digest algs from default, and gets into sign in the right context.

And that has exposed some holes in my provider -- it now imports peer's public key into "ovpn.xkey" and needs support for digest_verify functions. I was hoping to get away with private key related ops only. Although most of it is copy paste, this is already ~1000 lines and growing. All just to replace a few lines of code using hooks into EVP_PKEY_METHOD (or engine). 

A built-in shim provider that would require only private key ops implemented would greatly help application authors who have to support a variety of "external" keys.

**Comment by t8m** [comment]
I think a proper solution would be to have a way to indicate finer granularity of algorithm support to the libcrypto from a provider. I.E., libcrypto would know that your provider supports only sign but no verify operations and would use the default provider instead. Of course that is a possible future feature, not something that can and will be done in 3.0.

For now you can open a child library context inside your provider and simply forward the operations to the default provider via that context. Of course that would still require having and import function for the public key to be implemented and it is not a completely trivial workaround.

**Comment by selvanair** [comment]
One improvement could be to use the key's keymgmt to determine which provider's sign/verify ops should be called for it. Currently the signing context is determined by EVP_PKEY_CTX which is fine when the application creates it. With SSL_CTX created with some propq, it delegates all signature operation to the provider matching that propq irrespective of the key. When the key's keymgmt belongs to some other provider, this leads to the import/export process. And the need to implement DigestVerify which could have been directly handled by the internal default provider. Of course we only need to write some  glue for this,  but more code means more bugs.

Also allowing to leave out DigestSignInit and DigestSign to indicate that we prefer prefer SignInit and Sign would be nice. I thought that would be the case, but leaving them out leads to error.  This requires implementation of Digest in the provider as not all HSM's can do it internally (again, we can call OpenSSL back). Unless I missed some way of avoiding that.

**Comment by selvanair** [comment]
Thanks for the pointers and help. My provider is working and I can get a TLS session established. But when the provider is loaded groups secp256r1, secp512r1 and secp384r1 disappears from the list in extension 10. That breaks ECDH with RSA. I have RSA and EC implemented in keymgmt and query_keymgmt_name returns "RSA" or "ECDSA" for sign-operation, NULL otherwise. I haven't implemented keyexch assuming it will be resolved from default. 

The application is OpenVPN client, and without the provider all is fine. Is there any reason why this would happen? I'm loading using propq = "?provider=ovpn.xkey" and have default provider loaded. Normally we do not use SSL_CTX_set1_groups unless the user wants to override it. And explicitly setting something like prime256v1, just makes extension 10 disappear. So something is removing/forbidding those groups only when the provider is loaded.

**Comment by selvanair** [comment]
To add to that, here is what the client-hello has when traced from client side (the above was from what the server)

Using default provider only

```        
        extension_type=supported_groups(10), length=22
          ecdh_x25519 (29)
          secp256r1 (P-256) (23)
          ecdh_x448 (30)
          secp521r1 (P-521) (25)
          secp384r1 (P-384) (24)
          ffdhe2048 (256)
          ffdhe3072 (257)
          ffdhe4096 (258)
          ffdhe6144 (259)
          ffdhe8192 (260)
```
Using custom provider which overrides keymgmt and signature
```
        extension_type=supported_groups(10), length=22
          ecdh_x25519 (29)
          ecdh_x448 (30)
          ffdhe2048 (256)
          ffdhe3072 (257)
          ffdhe4096 (258)
          ffdhe6144 (259)
          ffdhe8192 (260)
```

**Comment by mattcaswell** [comment]
If your provider is supplying groups to be advertised in TLS then it needs to tell libssl about them using the "TLS-GROUPS" capability. See the "provider_get_capabilities" function and the "CAPABILITIES" section on this page:

https://github.com/openssl/openssl/blob/master/doc/man7/provider-base.pod

Also see how the default and fips providers handle this here:

https://github.com/openssl/openssl/blob/master/providers/common/capabilities.c

**Comment by selvanair** [comment]
I did a quick test by obtaining the list of capabilities in the default provider and returning that list. Not sure whether that's correct to do.

Anyway, that fixes ClientHello, server now responds with certificate which is fine and then ECDHE key exchange -- that fails:
```
error:03000096:digital envelope routines::operation not supported for this keytype
error:0A00013A:SSL routines::unable to find ecdh parameters
```
Looks like I need to implement something more. At the end of the day the only thing OpenSSL cannot access directly is the private key and the purpose of the provider is to interface with the backend and supply private key ops. Everything else, either I don't want my provider be called or I want to delegate back to the default provider.

**Comment by mattcaswell** [comment]
It looks clear to me that in 3.1 we need to be provide a better way of saying - its the same as that algorithm but with these differences, e.g. some kind of "filter" provider.

**Comment by selvanair** [comment]
I will get back to that "filter part" . Actually we only need a couple of groups by default and any custom ones that the user may provide as an option.

But it seems I have to implement EC key generation capability in the keymgmt. Kind of obvious in retrospect. I added a some empty code for gen_init, gen_set_params and gen and that is getting called. I was hoping that can avoid that... I can construct the EC key using default provider, so its not a big deal -- a lot more of glue code...

**Comment by selvanair** [comment]
> It looks clear to me that in 3.1 we need to be provide a better way of saying - its the same as that algorithm but with these differences, e.g. some kind of "filter" provider.

Obviously.. please spare us of the kind of unnecessary pain I'm going through. Its good to get used to the innards of provider interface though :)

**Comment by selvanair** [comment]
Now, how to delegate  EC key generation back to the default provider? Having never used keygen/paramgen, not sure what functions correspond to `keymgmt_gen_init`, `get_set_param` and `gen`. Here is what I tried as a shot in the dark, and that completes without error, but is it right?

In keymgmt_gen_init()
```
    EVP_PKEY_CTX *ctx = EVP_PKEY_CTX_new_from_name(prov->libctx, "EC", NULL);
    EVP_PKEY_CTX_set_params(ctx, p);
    EVP_PKEY_keygen_init(ctx);
```

The ctx is saved and passed on.

In keymgmt_gen_set_params()
```
EVP_PKEY_CTX_set_params(ctx, p)
```
Finally in keymgt generate
```
    EVP_PKEY *pkey = NULL;
    EVP_PKEY_generate(ctx, &pkey);
```

also need `set_template`, where I just used `EVP_PKEY_get_group_name` as I have the generated key hidden in the  keydata, and set it on the context using  EVP_PKEY_CTX_set_group_name. Is there anything else in the template?

At that point it queries keymgt name with OP_KEYEXCH where I return "ECDH". Then it wants keyexch implemented. I thought it will export the key out and use keyexch in default.

Anyway, I implemented keyexch using calls to `EVP_PKEY_derive` family of functions and all look good on surface. But the server rejects with "Verification of message MAC failed".

**Comment by mattcaswell** [comment]
Hmm. This latest difficulty with key gen and tls groups etc seems related to key exchange (ECDH) - not signatures (ECDSA). I'm wondering why you need to go into this - since your original problem was about signatures. Can you not use your own provider for signatures, and the default provider for key exchange?

**Comment by selvanair** [comment]
Yes, signature is the only thing I want to provide. SSL_CTX is taking ciphers, rand etc from default but it seems a whole lot of ops are tied to keyexch and there is no way to say call me only for signature.

Here is a summary of this thread:

If I create SSL_CTX in default context, then it wants to export the signing private key which is not possible. If I create SSL_CTX with "?provider=foo", then it always makes EVP_PKET_CTX in the context of foo's keymgmt, and all operations including peer's public key ops, EC key generation for ECDH  are being done through foo's keymgmt. 

Now, if I generate EC key is generated through my provider it queries it for OP_KEYEXCH and errors if I don't return "ECDH". Once I return that it looks for keyexch in my provider.

I'll be more than happy not having to intercept all those calls.

**Comment by mattcaswell** [comment]
Hmmm. This doesn't seem right. You should not have to implement KEYEXCH just because you have implemented SIGNATURE.

**Comment by selvanair** [comment]
> Hmmm. This doesn't seem right. You should not have to implement KEYEXCH just because you have implemented SIGNATURE.

I think this is what happens: the propq passed to SSL_CTX_new...() is used by it for all fetches. It finds everything from default except keymgmt and signature from me. But when it ceates EVP_PKEY_CTX for EC key generation (for ECDH), the context's keymgmt is my provider. So it asks me to generate the EC key.  If I can avoid that, the question about keyexch is moot.

**Comment by selvanair** [comment]
What is needed is a more fine-grained fetching. Like, if a keymgmt does not implement gen_init, gen_set_params and gen, take it from the next provider matching the propq. But it looks not possible in the current implementation of provider interface. Same with signature -- it should be possible to take DigestSign and Sign from one provider and DigestSignVerify and SignVerify from another.

**Comment by mattcaswell** [comment]
> Now, if I generate EC key through my provider it queries it for OP_KEYEXCH and errors if I don't return "ECDH". Once I return that it looks for keyexch in my provider. (I may be able to avoid keyexch by allowing to export the generated key -- I haven't tried as yet, as I'm stuck at proper key generation.)

This *should* be the case, i.e. if you are able to export the generated key, then it should export it and import it into some other provider that *is* able to do keyexch.

> I think this is what happens: the propq passed to SSL_CTX_new...() is used by it for all fetches. It finds everything from default except keymgmt and signature from me. But when it ceates EVP_PKEY_CTX for EC key generation (for ECDH), the context's keymgmt is my provider. So it asks me to generate the EC key. 

This sounds plausible. Hmm. I can't think of a way around this at the moment.

> What is needed is a more fine-grained fetching. Like, if a keymgmt does not implement gen_init, gen_set_params and gen, take it from the next provider matching the propq. 

Yes - agreed we do need to be able to do this. Additionally we should be able to do this on a "group" basis, i.e. if you want P-384 but the fetched keymgmt only does P-256 then keep looking until you find a keymgmt that can do this.

**Comment by selvanair** [comment]
Sounds like we should wait for 3.1 or 3.2 before we try to remove all deprecated API calls..

> > Now, if I generate EC key through my provider it queries it for OP_KEYEXCH and errors if I don't return "ECDH". Once I return that it looks for keyexch in my provider. (I may be able to avoid keyexch by allowing to export the generated key -- I haven't tried as yet, as I'm stuck at proper key generation.)
> 
> This _should_ be the case, i.e. if you are able to export the generated key, then it should export it and import it into some other provider that _is_ able to do keyexch.

I have a question on this. Once the key is generated it queries keymgmt_operation_name with id = OP_KEYEXCH (I know this because I made a bogus EC key for testing and tried to proceed). If I respond ECHD there, it expects to find keyexch with that name in the  the same provider. What else can I return here --- if I go by the docs, returning NULL would mean it would take the name to be same as that of the key ("EC") and expect the operation with that name implemented in the same provider.

So it seems keyexch is also tied to keymgmt?

In any case, hoping that the granular approach you mention gets implemented in future versions.

**Comment by mattcaswell** [comment]
> it expects to find keyexch with that name in the the same provider.

This is actually quite surprising. I would expect it to export the key automatically to the right provider. I think there is a bug somewhere here....need to look, but I won't get time today.

**Comment by selvanair** [comment]
My understanding of OpenSSL code base is poor to put it mildly, so take all of the below with a grain of salt (I stepped through it in debugger to confirm, though).

For ECDHE, `SSL_derive()` calls `EVP_PKEY_derive_init()` with a context created with the generated EC key and propq of SSL_CTX. So, this `EVP_PKEY_CTX` will have my provider's keymgmt in it. I refer to this context as `ctx` below.

Now, in `EVP_PKEY_dervive_ex_init()` in `crypto/ev/exchange.c` where that call lands, an export is attempted (line 233), but that's to the keymgmt of the `ctx`. Which is a no-op as the key is from there. At this point it doesn't know where exchange ops (keyexch) are going to come from. It does a `query_operation_name` (line 246) on the ctx->keymgmt which lands in my provider an I return "ECDH". 

That seems to be fine as it still fetches keyexch using that name "ECDH" and finds it in default provider (see around line 260 of `exchange.c`)

So far so good except for the fact that the key is still in my provider and exchange ops are in default. Then there is a check whether the provider of exchange matches that of the `ctx->keymgmt`. That obviously fails. 

```
260         exchange = EVP_KEYEXCH_fetch(ctx->libctx, supported_exch, ctx->propquery);
```

It finds exchange in default provider as I don't implement it.

```
261
262         if (exchange == NULL
263             || (EVP_KEYMGMT_get0_provider(ctx->keymgmt)
264                 != EVP_KEYEXCH_get0_provider(exchange))) {
265             /*
266              * We don't need to free ctx->keymgmt here, as it's not necessarily
267              * tied to this operation.  It will be freed by EVP_PKEY_CTX_free().
268              */
269             EVP_KEYEXCH_free(exchange);
270             goto legacy;
271         }
```

It goes to legacy and errors out as the key is not legacy.

It looks like a design decision to always export to the provider of `ctx->keymgmt` instead of to the provider of the operation.

**Comment by t8m** [comment]
Hmm, there is actually no way to fetch a keymgmt associated with the provider from the exchange method. IMO we need that function.

**Comment by t8m** [comment]
> It looks like a design decision to always export to the provider of `ctx->keymgmt` instead of to the provider of the operation.

I would not call it a design decision, I'd call it a mistake. If I am able to fetch the operation why not export the data to the provider of the operation.

**Comment by levitte** [comment]
> > it expects to find keyexch with that name in the the same provider.
> 
> This is actually quite surprising. I would expect it to export the key automatically to the right provider. I think there is a bug somewhere here....need to look, but I won't get time today.

This is from the original comment up top:

> In this case, all is good until signature is needed which fails with calls to provider's keymgmt_export() with PRIVATE_KEY in selection, not possible for the key. 

That surely impacts on the possibility to "export the key automatically to the right provider"

**Comment by t8m** [comment]
No, the issue described in the comment above (https://github.com/openssl/openssl/issues/16614#issuecomment-929714081) is a different case. Here the key is a temporary ECDH key exchange key so it should be exportable.

**Comment by mattcaswell** [comment]
I'm working on a solution to this.

**Comment by t8m** [comment]
@mattcaswell I've started with https://github.com/t8m/openssl/commit/172da3757ce9a39b4c4d054d68135ec19eb7c6a7 which is of course wrong as I've commented - the missing step is to get a keymgmt associated with the operation's provider.

**Comment by selvanair** [comment]
While exporting to the operation's provider would generally help (and may be the right thing to do as you say), I wish for a deeper solution: 

If the custom provider does not implement key generation functions (gen_init, gen_set_params and gen), take it from the next one in line. Same with groups. That would make it so much easier for a provider author wanting to support signing with a non-exportable key.  This may require non-trivial changes to the way fetch works, so not to be expected in near-term?

Anyway, we'll have to hold-off on our plan to use a provider for external keys at least until 3.1. We would rather wait if a more "granular" approach could be expected in, say, 3.2.

(Should I open a new issue with this request?)

**Comment by mattcaswell** [comment]
> @mattcaswell I've started with t8m@172da37 which is of course wrong as I've commented - the missing step is to get a keymgmt associated with the operation's provider.

I've got something not dissimilar to this which I think will work. Although I'm currently missing a test to confirm it.

**Comment by mattcaswell** [comment]
I think there are two problems that I think are probably bugs in 3.0 which we probably can fix:

1) If we have an EVP_PKEY_CTX with a pkey in it from a given provider, and the keymgmt for that pkey does not support exporting the key, then any fetches associated with that EVP_PKEY_CTX should not return an algorithm that is in anything other than the same provider.

2) Having fetched an algorithm for use in an EVP_PKEY_CTX (e.g. a KEYEXCH), then we should always ensure that the keymgmt is in the same provider, and update the keymgmt in the EVP_PKEY_CTX if it is not.

These two are subtly different, but closely related.

**Comment by mattcaswell** [comment]
If we fix these two things then I think it goes a long way to solving @selvanair's problems (you would not need to implement keygen, or anything to do with keyexch, or tls capabilities). You probably would still need to handle the keymgmt_operation_name with id = OP_KEYEXCH thing though.

Basically if we fixed those bugs then you could fetch the key as now:

````
EVP_PKEY *pkey = my_custom_loader(libctx, "provider=ovpn.xkey", uri); /* this loads an opaque pointer as keydata */ 
````
And then setup SSL_CTX to prefer the default provider over yours:

````
SSL_CTX *ctx = SSL_CTX_new_ex(libctx, "?provider=default", method);
SSL_CTX_use_certificate(ctx, cert);
SSL_CTX_use_Privatekey(ctx, pkey);
````
Doing it this way, libssl would always use algorithms from default whenever possible. However the private key you loaded would have a keymgmt from your customer provider, and (with the bugs mentioned above fixed) then, because your provider cannot export, it would only fetch a SIGNATURE algorithm from your provider (overriding the preference in the propquery for the default provider). I think this even works for the verify stuff, i.e. when verifying you would not be using a key from your custom provider and therefore the SIGNATURE algorithm fetched would be from the default provider and therefore you should be able to get away without having to implement the verify code in your provider.

**Comment by mattcaswell** [comment]
Ok, I have something that you might like to give a try. See #16711.

This is a WIP. It compiles and doesn't break any regression tests, but I currently don't have a test to check that it has the desired effect. It would be great if you can give it a go and let me know how you get on.

To get this to work you should set things up as I previously described, i.e. load the key using your custom provider, but when setting up the SSL_CTX make sure you prefer the default provider (property query: "?provider=default"). This should hopefully mean the default provider is used in preference for all operations, and only use your provider when it has to use your key.

**Comment by selvanair** [comment]
I did a quick test with "?provider=default" and no export() defined in keymgmt. 

Something is not quite right with the way propq is re-written -- I think the issue is that when not exportable, the propq becomes "?provider=default,provider=ovpn.xkey" and that returns NULL for signature. Is such combinations supported?

Even with a normal key in a pem file directly set into SSL_CTX, connection fails to verify even peer's signature if my provider is loaded. That starts to work if I use libctx = NULL in my provider, though. In this case possibly there are two default providers -- one in NULL and one in my libctx which I load. Then key and signature could come from different providers though the key is exportable. In that case signature is refetched using "?provider=default,provider=default" and that too fails to find a signature operation. 

Even if comma separated  words work, is it not required to just use the exact match with "provider=foo" in this case to be sure to fetch the right operation?

Is the use of multiple specs in propq, with or without "?" etc documented somewhere? I could never find a satisfactory documentation of this. Pointer to comments in the codes or commit message would do if there are no docs.

**Comment by paulidale** [comment]
> Something is not quite right with the way propq is re-written -- I think the issue is that when not exportable, the propq becomes "?provider=default,provider=ovpn.xkey" and that returns NULL for signature. Is such combinations supported?

Not really.  The query code isn't expecting the same name in a query twice.  Whichever is second (which is undefined) will always fail to match.

Even though it's not something we should be creating, I think an error should be raised when parsing such a query.
Property queries can be merged safely -- there is an internal call to do this.  They should never be concatenated as strings.

**Comment by selvanair** [comment]
Okay, thanks. Editing t he patch to just return "property=foo", it does seem to work. I need to do more tests though.

**Comment by mattcaswell** [comment]
> Not really. The query code isn't expecting the same name in a query twice. Whichever is second (which is undefined) will always fail to match.

Hmm. That's unexpected. I was expecting a propquery that said "?provider=default,provider=ovpn.xkey", to apply both restrictions. The first one says "prefer the default provider". The second one says "always use the ovp.xkey provider". So I was expecting it to effectively ignore the preference since there will never be an algorithm that matches the second clause that also comes from the default provider.

**Comment by paulidale** [comment]
Setting the default query to `?property=default` and passing a query of `provider=ovpn.xkey` does work as expected.

Setting both in one query, the order matter and one will apply.  I don't think we need to rely on this.

**Comment by levitte** [comment]
> If the custom provider does not implement key generation functions (gen_init, gen_set_params and gen), take it from the next one in line. Same with groups. That would make it so much easier for a provider author wanting to support signing with a non-exportable key. This may require non-trivial changes to the way fetch works, so not to be expected in near-term?

Unfortunately, we do not have the concept of "take it from the next one in line".  This was a design decision made very early (all the way back 3 years ago, when we started talking about the provider functionality), that a fetch simply gets you back one method, and if there are several choices, you get whatever comes first.

In retrospect, that design decision turns out to be a bit too blunt.
To be noted, thought, that there are "do_all" methods that iterate through all the methods given certain search criteria, called `EVP_{NAME}_do_all_provided()` (for example, `EVP_SIGNATURE_do_all_provided()`), which could be used for that sort of purpose.  It comes with a lot of churn, though...

**Comment by levitte** [comment]
Somewhere in the end, this appears to be a conflict of what the caller actually wants, and this comes all the way down to where operations are initialized.

So far, the priority has been that whatever the caller asks for rules.  If the caller has set up a key in one provider and has asked for the operation to happen in another provider, then that rules, and an export of the key to the operation provider *must* happen.

This does get complicated with the SSL library, where we may get a mix of different asks through different SSL functions.  I would rather think that the fix should be there, rather than making the EVP functionality less precise.

**Comment by mattcaswell** [comment]
> Somewhere in the end, this appears to be a conflict of what the caller actually wants, and this comes all the way down to where operations are initialized.

I simply don't see what the conflict is?

Load a key from provider "ovpn.xkey" and tell your SSL_CTX to use it. Also tell your SSL_CTX to *prefer* the default provider (propquery "?provider=default"). Where is the conflict? I expect algorithms to always be fetched from the default provider, except in the case where we need to use the (unexportable) key. In that case our preference to use the default provider doesn't apply because we *cannot* use the default provider.

> So far, the priority has been that whatever the caller asks for rules

It is perfectly possible for EVP to respect what the caller has asked for, and still be able to operate without failing in the above scenario. The fact that it currently fails is clearly a *bug* in my eyes.

> I would rather think that the fix should be there, rather than making the EVP functionality less precise.

I fail to see what is making EVP less precise here.

**Comment by selvanair** [comment]
What is the actual purpose/meaning of no_store in provider query_operation()? Its value appears to affect which method is fetched -- this is with git master.

**Comment by selvanair** [comment]
> For now you can open a child library context inside your provider and simply forward the operations to the default provider via that context. 

Regarding the use of a child library context: what is the difference between creating a new child `libctx` vs just using the `libctx` with which the provider init is called ? --  one can get a  pointer to it in provider init by calling `OSSL_FUNC_core_get_libctx(in)` The latter will have all providers loaded by the application into it, including future ones, isn't it? In the documentation the meaning of "application libctx" is unclear -- I suppose an application may have multiple libctx's and what is referred to here is the one used to init the provider, right?

Also, if the application prefers some other provider than the `"default"` for most operations, how to respect that from a provider that wants to forward some tasks out? What I do right now is to use a `propq` like `"provider!=foo"` where `foo` is my provider to avoid the call looping back to me. But that doesn't look ideal.
